crates/net/network/src/transactions.rs

```diff
@@ -1074,116 +1188,266 @@ struct Peer {
 ///
 /// This will keep track of unique transaction hashes that are currently being fetched and submits
 /// new requests on announced hashes.
-#[derive(Debug, Default)]
+#[derive(Debug)]
+#[pin_project]
 struct TransactionFetcher {
+    /// All peers to which a request for pooled transactions is currently active. Maps 1-1 to
+    /// `inflight_requests`.
+    active_peers: LruMap<PeerId, u8>,
     /// All currently active requests for pooled transactions.
+    #[pin]
     inflight_requests: FuturesUnordered<GetPooledTxRequestFut>,
-    /// Set that tracks all hashes that are currently being fetched.
-    inflight_hash_to_fallback_peers: HashMap<TxHash, Vec<PeerId>>,
+    /// Hashes that are awaiting fetch from an idle peer.
+    buffered_hashes: LruCache<TxHash>,
+    /// Queue of hashes that have been evicted from the hash buffer.
+    dropped_hashes_rx: UnboundedReceiver<TxHash>,
```

Comment:
if I understood this correctly, then this receives the evicted hashes from `buffered_hashes`.

`buffered_hashes` ensures that future work is bound and we have a fixed capacity of pending hashes that we need to fetch.

though, since `TransactionFetcher` owns the buffer itself, this channel roundtrip looks redundant to me, because this is always fired on insert which we control.
So perhaps we add a new insert function to the cache that returns the evicted value (Option<T>)